File: binary-tree-level-order-traversal/binary-tree-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
import collections
class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        ans = []

        if root is None:
            return ans

        queue = collections.deque()
        queue.append(root)

        while queue:
            currSize = len(queue)
            currList = []

            while currSize > 0:
                currNode = queue.popleft()
                currList.append(currNode.val)
                currSize -= 1

                if currNode.left is not None:
                    queue.append(currNode.left)
                if currNode.right is not None:
                    queue.append(currNode.right)
            ans.append(currList)

        return ans

File: binary-tree-level-order-traversal/binary-tree-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    vector<vector<int>> levelOrder(TreeNode *root)
    {
        vector<vector<int>> ans;
        if (!root)
            return ans;

        // initialise queue
        queue<TreeNode *> q;
        q.push(root);

        while (!q.empty())
        {
            int currSize = q.size();
            vector<int> currList;
            while (currSize > 0)
            {
                // Dequeue element
                TreeNode *currNode = new TreeNode();
                currNode = q.front();

                currList.push_back(currNode->val);
                currSize -= 1;
                q.pop();

                //Check for left child
                if (currNode->left)
                    q.push(currNode->left);
                //Check for right child
                if (currNode->right)
                    q.push(currNode->right);
            }
            ans.push_back(currList);
        }

        return ans;
    }
};